api/lib/domain/services/certification-challenges-service.js

```diff
@@ -1,19 +1,93 @@
 const _ = require('lodash');
 const CertificationChallenge = require('../models/CertificationChallenge');
+const {
+  MAX_CHALLENGES_PER_SKILL_FOR_CERTIFICATION,
+} = require('../constants');
+
+const KnowledgeElement = require('../models/KnowledgeElement');
+const UserCompetence = require('../models/UserCompetence');
+const Challenge = require('../models/Challenge');
+const challengeRepository = require('../../infrastructure/repositories/challenge-repository');
+const answerRepository = require('../../infrastructure/repositories/answer-repository');
+const knowledgeElementRepository = require('../../infrastructure/repositories/knowledge-element-repository');
 
 module.exports = {
 
-  generateCertificationChallenges(userCompetences, certificationCourseId) {
-    return _.flatMap(userCompetences, (userCompetence) => {
-      return _.map(userCompetence.challenges, (challenge) => {
-        return new CertificationChallenge({
-          challengeId: challenge.id,
-          competenceId: challenge.competenceId,
-          associatedSkillName: challenge.testedSkill.name,
-          associatedSkillId: challenge.testedSkill.id,
-          courseId: certificationCourseId,
-        });
+  async pickCertificationChallenges(placementProfile) {
+    const knowledgeElementsByCompetence = await knowledgeElementRepository
+      .findUniqByUserIdGroupedByCompetenceId({ userId: placementProfile.userId, limitDate: placementProfile.profileDate });
+
+    const knowledgeElements = KnowledgeElement.findDirectlyValidatedFromGroups(knowledgeElementsByCompetence);
+    const answerIds = _.map(knowledgeElements, 'answerId');
+
+    const challengeIdsCorrectlyAnswered = await answerRepository.findChallengeIdsFromAnswerIds(answerIds);
+
+    const allChallenges = await challengeRepository.findOperative();
+    const challengesAlreadyAnswered = challengeIdsCorrectlyAnswered.map((challengeId) => Challenge.findById(allChallenges, challengeId));
```

Comment:
Je vous laisse juger de décider si cet appel à la statique a un grand intérêt de lisibilité plutôt que :
```suggestion
    const challengesAlreadyAnswered = challengeIdsCorrectlyAnswered.map((challengeId) => allChallenges.find((challenge) => challenge.id === challengeId));
```